
# Практическая работа 4-2: Линейная регрессия 2

Name: Егоров А.С.

Email: Egorov_AS@mgpu.ru

## ##  Анализ набора кредитных данных

Вспомните набор данных «Кредит», рассмотренный на занятии и доступный [здесь](http://www-bcf.usc.edu/~gareth/ISL/data.html).
Этот набор данных состоит из информации о кредитных картах 400 человек.

Сначала импортируйте данные и преобразуйте доход в тысячи.

In [1]:
# Импорт данных в Google Colab
from google.colab import files
uploaded = files.upload()

Saving Credit.csv to Credit.csv


In [2]:
# imports and setup

import scipy as sc
import numpy as np

import pandas as pd
import statsmodels.formula.api as sm     #Last lecture: used statsmodels.formula.api.ols() for OLS
from sklearn import linear_model         #Last lecture: used sklearn.linear_model.LinearRegression() for OLS

import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (10, 6)

from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm

# Import data from Credit.csv file
credit = pd.read_csv('Credit.csv',index_col=0) #load data
credit["Income"] = credit["Income"].map(lambda x: 1000*x)
credit

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1     14891.0   3606     283      2   34         11    Male      No     Yes   
2    106025.0   6645     483      3   82         15  Female     Yes     Yes   
3    104593.0   7075     514      4   71         11    Male      No      No   
4    148924.0   9504     681      3   36         11  Female      No      No   
5     55882.0   4897     357      2   68         16    Male      No     Yes   
..        ...    ...     ...    ...  ...        ...     ...     ...     ...   
396   12096.0   4100     307      3   32         13    Male      No     Yes   
397   13364.0   3838     296      5   65         17    Male      No      No   
398   57872.0   4171     321      5   67         12  Female      No     Yes   
399   37728.0   2525     192      1   44         13    Male      No     Yes   
400   18701.0   5524     415      5   64          7  Female      No      No   

            Ethnicity  Balance  
1           Caucasian      333  
2               Asian      903  
3               Asian      580  
4               Asian      964  
5           Caucasian      331  
..                ...      ...  
396         Caucasian      560  
397  African American      480  
398         Caucasian      138  
399         Caucasian        0  
400             Asian      966  

[400 rows x 11 columns]

## Задание 1. Первая регрессионная модель

**Упражнение:** Лимит средств по рейтингу:


$$
\text{Limit} = \beta_0 + \beta_1 \text{Rating}.
$$

Поскольку кредитные рейтинги в основном используются банками для определения кредитных лимитов, ожидаем, что рейтинг будет в прямой зависимости для лимита, поэтому эта регрессия должна быть очень хорошей.

Используйте функцию `ols` из библиотеки Python `statsmodels`. Каково значение $R^2$? Что такое $H_0$ и $H_A$ для соответствующей проверки гипотезы и каково значение $p$?


In [4]:
# Ипользуем функцию ols из библиотеки Python statsmodels, чтобы найти переменные β0 и β1, R2
m = sm.ols(formula="Limit ~ Rating", data=credit).fit()
print(m.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.994
Model:                            OLS   Adj. R-squared:                  0.994
Method:                 Least Squares   F-statistic:                 6.348e+04
Date:                Mon, 13 May 2024   Prob (F-statistic):               0.00
Time:                        16:08:55   Log-Likelihood:                -2649.1
No. Observations:                 400   AIC:                             5302.
Df Residuals:                     398   BIC:                             5310.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -542.9282     22.850    -23.760      0.0

**Ваш ответ:**
β0=-542,93

β1=14,87

Уравнение получается: Limit = -542.93 + 14.87*Rating

R2 = 0.994 означает,что модель объясняет 99.4% изменчивости кредитного лимита.
Значение p меньше 0.05, значит, это указывает на сильную линейную связь и подтверждает значимость модели. Лимит(Limit) прямо зависит от рейтинга(Rating).


In [27]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Rating', y='Limit', title='Зависимость лимитов от кредитного рейтинга',)
fig.show()

График подтверждает прямую зависимость лимита кредита от кредитного рейтинга

## Задание 2: Прогнозирование лимита без рейтинга

Поскольку `Рейтинг` и `Лимит` — это практически одна и та же переменная, исключим `Рейтинг` и просто попробуем спрогнозировать `Лимит` на основе вещественнозначных переменных (некатегориальных переменных): Доход, Карты, Возраст, Образование, Баланс.

**Упражнение.** Разработайте модель множественной линейной регрессии для прогнозирования рейтинга. Интерпретируйте результаты.

А пока просто сосредоточьтесь на переменных с численными значениями (доход, карты, возраст, образование, баланс).

В первом приближении не использовать категориальные переменные (Пол, Студент, Семейное положение, Этническая принадлежность).



Ипользуем функцию ols из библиотеки Python statsmodels и построим модели с вещественными переменными

In [28]:
# Зависимость Лимита от Дохода
m2 = sm.ols(formula = "Limit ~ Income", data = credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.627
Model:                            OLS   Adj. R-squared:                  0.626
Method:                 Least Squares   F-statistic:                     670.2
Date:                Mon, 13 May 2024   Prob (F-statistic):           2.39e-87
Time:                        18:10:33   Log-Likelihood:                -3467.3
No. Observations:                 400   AIC:                             6939.
Df Residuals:                     398   BIC:                             6947.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2389.8694    114.829     20.812      0.0

P-значение маленькое (меньше 0,05), R2 = 0.627 - 62,7% изменчивости крединого лимита

In [29]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Income', y='Limit', title='Зависимость лимитов от дохода заемщиков',)
fig.show()

На графике наблюдается зависимость лимита от доходов заемщика.

In [8]:
# Зависимость Лимита от Cards
m2 = sm.ols(formula = "Limit ~ Cards", data = credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                   0.04167
Date:                Mon, 13 May 2024   Prob (F-statistic):              0.838
Time:                        16:31:31   Log-Likelihood:                -3664.7
No. Observations:                 400   AIC:                             7333.
Df Residuals:                     398   BIC:                             7341.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4684.6662    274.978     17.037      0.0

p-значение=0,838, что больше 0,05. R2 = 0. Между данными перменными связи нет, модель не значима

In [30]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Cards', y='Limit', title='Зависимость лимитов от Cards',)
fig.show()

На графике видно, что отсутствует зависимость

In [9]:
# Зависимость Лимита от возраста заемщика
m2 = sm.ols(formula = "Limit ~ Age", data = credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.010
Model:                            OLS   Adj. R-squared:                  0.008
Method:                 Least Squares   F-statistic:                     4.093
Date:                Mon, 13 May 2024   Prob (F-statistic):             0.0437
Time:                        16:31:47   Log-Likelihood:                -3662.7
No. Observations:                 400   AIC:                             7329.
Df Residuals:                     398   BIC:                             7337.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3984.0985    388.857     10.246      0.0

p-значение = 0,044 (меньше 0,05, но близко к нему), R2=0.01, то есть 1% изменчивости кредитного лимита. Модель не значима

In [32]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Age', y='Limit', title='Зависимость лимитов от возраста заемщика',)
fig.show()

Возраст заемщика практически не влияет на кредитный лимит, основываясь графика

In [33]:
# Зависимость Лимита от образования заемщика
m2 = sm.ols(formula = "Limit ~ Education", data = credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.2208
Date:                Mon, 13 May 2024   Prob (F-statistic):              0.639
Time:                        18:16:11   Log-Likelihood:                -3664.7
No. Observations:                 400   AIC:                             7333.
Df Residuals:                     398   BIC:                             7341.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4969.5271    511.029      9.725      0.0

p-значение = 0,639 (больше 0,05), R2=0.001, то есть 0,1% изменчивости кредитного лимита. Модель не значима, связь отсутствует между перемнными

In [34]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Education', y='Limit', title='Зависимость лимитов от образования заемщика',)
fig.show()

На графике видно, что отсутствует зависимость

In [35]:
# Зависимость Лимита от Balance
m2 = sm.ols(formula = "Limit ~ Balance", data = credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.743
Model:                            OLS   Adj. R-squared:                  0.742
Method:                 Least Squares   F-statistic:                     1148.
Date:                Mon, 13 May 2024   Prob (F-statistic):          2.53e-119
Time:                        18:17:27   Log-Likelihood:                -3393.4
No. Observations:                 400   AIC:                             6791.
Df Residuals:                     398   BIC:                             6799.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   2485.9566     88.586     28.063      0.0

p-значение меньше 0,05, R2=0.743, то есть 74,3% изменчивости кредитного лимита.

In [36]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Balance', y='Limit', title='Зависимость лимитов от Balance',)
fig.show()

На основе графика можно сделать вывод, что кредитный лимит имеет зависимость от Balance.

На основе запросов выше, можно сказать, что переменные Income и Balance являются хорошими предикторами, так как у них очень маленькое p-значение, которое намного меньше 0.05 и хорошее соответствие модели данным.

In [19]:
top_m2 = sm.ols(formula="Limit ~ Income + Balance", data = credit).fit()
print(top_m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.939
Model:                            OLS   Adj. R-squared:                  0.939
Method:                 Least Squares   F-statistic:                     3046.
Date:                Mon, 13 May 2024   Prob (F-statistic):          1.37e-241
Time:                        17:03:08   Log-Likelihood:                -3106.0
No. Observations:                 400   AIC:                             6218.
Df Residuals:                     397   BIC:                             6230.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1610.3665     49.710     32.395      0.0

Какие независимые переменные являются хорошими/плохими предикторами? Какая модель в целом является лучшей?

**Ваш ответ:**
Плохие предиторы: Cards, Age, Education

Хорошие предиторы: Income и Balance.

На основе построенной модели из Income и Balance, получаем, что

Limit=β0+β1∗Income+β2∗Balance

β0=1610.37

β1=0.03

β2=3.16

Limit=1610.37+0.03∗Income+3.16∗Balance

R2 = 0.939 означает,что модель объясняет  93.9%  изменчивости кредитного лимита. Это очень высокое значение, данные соответствуют модели.

p-значение меньше 0.05, значит, регрессия имеет сильную линейную связь и подтверждает значимость модели.


In [37]:
# Проверим, что будет, если включить все перемнные модель, как изменится значение R2
m2 = sm.ols(formula="Limit ~ Income + Cards + Age + Education + Balance", data=credit).fit()
print(m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.940
Model:                            OLS   Adj. R-squared:                  0.940
Method:                 Least Squares   F-statistic:                     1242.
Date:                Mon, 13 May 2024   Prob (F-statistic):          1.32e-238
Time:                        18:20:22   Log-Likelihood:                -3101.0
No. Observations:                 400   AIC:                             6214.
Df Residuals:                     394   BIC:                             6238.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   1752.3883    170.621     10.271      0.0

Как мы видим R2=0,94 - 94% изменчивости кредитного лимита. Это на 0,1 больше, чем у прошлой модели. Делаем вывод, наилучшей моделью признана зависимость кредитного лимта как от Balance, так и от Income

## Задание 3. Включение категориальных переменных в модель регрессии

Рассмотреть двоичные категориальные переменные, которые были ранее сопоставлены с целочисленными значениями 0, 1 в в `лекции 4.2`.

In [14]:
credit["Gender_num"] = credit["Gender"].map({' Male':0, 'Female':1})
credit["Student_num"] = credit["Student"].map({'Yes':1, 'No':0})
credit["Married_num"] = credit["Married"].map({'Yes':1, 'No':0})
credit

Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
1     14891.0   3606     283      2   34         11    Male      No     Yes   
2    106025.0   6645     483      3   82         15  Female     Yes     Yes   
3    104593.0   7075     514      4   71         11    Male      No      No   
4    148924.0   9504     681      3   36         11  Female      No      No   
5     55882.0   4897     357      2   68         16    Male      No     Yes   
..        ...    ...     ...    ...  ...        ...     ...     ...     ...   
396   12096.0   4100     307      3   32         13    Male      No     Yes   
397   13364.0   3838     296      5   65         17    Male      No      No   
398   57872.0   4171     321      5   67         12  Female      No     Yes   
399   37728.0   2525     192      1   44         13    Male      No     Yes   
400   18701.0   5524     415      5   64          7  Female      No      No   

            Ethnicity  Balance  Gender_num  Student_num  Married_num  
1           Caucasian      333           0            0            1  
2               Asian      903           1            1            1  
3               Asian      580           0            0            0  
4               Asian      964           1            0            0  
5           Caucasian      331           0            0            1  
..                ...      ...         ...          ...          ...  
396         Caucasian      560           0            0            1  
397  African American      480           0            0            0  
398         Caucasian      138           1            0            1  
399         Caucasian        0           0            0            1  
400             Asian      966           1            0            0  

[400 rows x 14 columns]

Можете ли улучшить модель, которая разработана в задании `2`, включив в нее одну или несколько из этих переменных?


Проверим каждую новую переменную, ипользуя функцию ols из библиотеки Python statsmodels

In [15]:
new_top_m2 = sm.ols(formula="Income ~ Gender_num", data=credit).fit()
print(new_top_m2.summary())

                            OLS Regression Results                            
Dep. Variable:                 Income   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                   0.04589
Date:                Mon, 13 May 2024   Prob (F-statistic):              0.830
Time:                        16:55:52   Log-Likelihood:                -4755.1
No. Observations:                 400   AIC:                             9514.
Df Residuals:                     398   BIC:                             9522.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   4.561e+04   2539.977     17.957      0.0

p-значение = 0,83, что больше 0,05, R2 = 0. Между данными перменными связи нет, модель не значима

In [38]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Gender_num', y='Limit', title='Зависимость лимитов от Gender_num',)
fig.show()

На графике видно, что зависимость отсутствует

In [16]:
new_top_m2 = sm.ols(formula="Income ~ Student_num", data=credit).fit()
print(new_top_m2.summary())

                            OLS Regression Results                            
Dep. Variable:                 Income   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1535
Date:                Mon, 13 May 2024   Prob (F-statistic):              0.695
Time:                        16:56:17   Log-Likelihood:                -4755.0
No. Observations:                 400   AIC:                             9514.
Df Residuals:                     398   BIC:                             9522.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    4.499e+04   1859.510     24.194      

p-значение = 0,695, что больше 0,05, R2 = 0. Между данными перменными связи нет, модель не значима

In [39]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Student_num', y='Limit', title='Зависимость лимитов от Student_num',)
fig.show()

На графике видно, что нет никакой зависимости

In [17]:
new_top_m2 = sm.ols(formula="Income ~ Married_num", data=credit).fit()
print(new_top_m2.summary())

                            OLS Regression Results                            
Dep. Variable:                 Income   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.5065
Date:                Mon, 13 May 2024   Prob (F-statistic):              0.477
Time:                        16:56:46   Log-Likelihood:                -4754.8
No. Observations:                 400   AIC:                             9514.
Df Residuals:                     398   BIC:                             9522.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    4.364e+04   2832.640     15.407      

p-значение = 0,077 (меньше 0,05, но близко к нему), R2=0.001, то есть 0,1% изменчивости кредитного лимита

In [40]:
import matplotlib.pyplot as plt
import plotly.express as px
fig = px.scatter(credit, x='Married_num', y='Limit', title='Зависимость лимитов от Married_num',)
fig.show()

На графике видно, что нет никакой зависимости

In [43]:
# Проверим, зависит ли кредитный лимит от прошлой модели плюс наиболее подходящая новая перменная Married_num
new_top_m2 = sm.ols(formula="Limit ~ Income + Balance + Student_num", data=credit).fit()
print(new_top_m2.summary())

                            OLS Regression Results                            
Dep. Variable:                  Limit   R-squared:                       0.974
Model:                            OLS   Adj. R-squared:                  0.973
Method:                 Least Squares   F-statistic:                     4875.
Date:                Mon, 13 May 2024   Prob (F-statistic):          3.55e-312
Time:                        18:32:50   Log-Likelihood:                -2937.6
No. Observations:                 400   AIC:                             5883.
Df Residuals:                     396   BIC:                             5899.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    1671.8940     32.784     50.998      

**Ваш ответ:**
На основе проведенного анализа, можно сделать вывод, что ни одна из новых перемнных не явяется значимой и не влияет на кредитный лимит студента по отдельности. Однако перемнная Married_num с перемнными Balance и Income вместе дают R2=0,974, что выше предыдущего рекорда на 3,4% и эта модель теперь является лучшей